src/box/lua/config/configdata.lua

```diff
@@ -246,6 +253,167 @@ local function effective_snapshot_dir(iconfig)
     return fio.abspath(fio.pathjoin(work_dir, snap_dir))
 end
 
+-- Read snap file and return a map of saved UUIDs and names for
+-- all instances and for the current repicaset.
+local function snapshot_names(snap_path)
+    local peers = {}
+    local instance_uuid = xlog.meta(snap_path).instance_uuid
+    local instance_name, replicaset_name, replicaset_uuid
+    for _, row in xlog.pairs(snap_path) do
+        local body = row.BODY
+        if not body.space_id then
+            goto continue
+        end
+
+        if body.space_id > box.schema.CLUSTER_ID then
+            -- No sense in scanning after _cluster.
+            break
+        end
+
+        if body.space_id == box.schema.SCHEMA_ID then
+            if body.tuple[1] == 'replicaset_uuid' then
+                replicaset_uuid = body.tuple[2]
+            elseif body.tuple[1] == 'replicaset_name' then
+                replicaset_name = body.tuple[2]
+            end
+        elseif body.space_id == box.schema.CLUSTER_ID then
+            if body.tuple[2] == instance_uuid then
+                instance_name = body.tuple[3]
+            end
+
+            table.insert(peers, {
+                instance_uuid = body.tuple[2],
+                instance_name = body.tuple[3],
+            })
+        end
+        ::continue::
+    end
+
+    return {
+        replicaset_name = replicaset_name,
+        replicaset_uuid = replicaset_uuid,
+        instance_name = instance_name,
+        instance_uuid = instance_uuid,
+        peers = peers,
+    }
+end
+
+local function find_peer_name_by_uuid(peers, instance_uuid)
+    for name, peer in pairs(peers) do
+        local uuid = instance_config:get(peer.iconfig_def,
+            'database.instance_uuid')
+        if uuid == instance_uuid then
+            return name
+        end
+    end
+    return nil
+end
+
+-- Validate UUIDs and names passed to config against the data,
+-- saved inside snapshot. Fail early and don't recover all
+-- data if mismatch is found. Returns a map of names, missing
+-- in snapshot in the following format: {<name> = <uuid>}
+local function validate_names(snap_names, config_names)
+    -- Snapshot always has replicaset uuid and at least one
+    -- peer in _cluster space.
+    assert(snap_names.replicaset_uuid)
+    assert(snap_names.instance_uuid)
+    assert(#snap_names.peers > 0)
+    -- Config always has names set.
+    assert(config_names.replicaset_name ~= nil)
+    assert(config_names.instance_name ~= nil)
+
+    if config_names.replicaset_uuid ~= nil and
+       config_names.replicaset_uuid ~= snap_names.replicaset_uuid then
+        error(string.format('Replicaset UUID mismatch. Snapshot: %s, ' ..
+                            'config: %s.', snap_names.replicaset_uuid,
+                            config_names.replicaset_uuid))
```

Comment:
Nit: I would use `error(<...>, 0)` to don't show internal file names to a user in stderr on a startup error or in alerts on reload. It would looks like an internal logic error, so, as for me, it is better to just give an error message.